chat-client/src/client/mynahUi.ts

```diff
@@ -836,7 +857,106 @@ export const createMynahUi = (
         })
     }
 
+    /**
+     * Adjusts the UI when the user changes to/from free-tier/paid-tier.
+     * Shows a message if the user reaches free-tier limit.
+     * Shows a message if the user just upgraded to paid-tier.
+     */
+    const onPaidTierModeChange = (tabId: string, mode: string | undefined) => {
+        if (!mode || !['freetier', 'freetier-limit', 'upgrade-pending', 'paidtier'].includes(mode)) {
+            return false // invalid mode
+        }
+
+        tabId = !!tabId ? tabId : getOrCreateTabId()!
+        const store = mynahUi.getTabData(tabId).getStore() || {}
+
+        // Detect if the tab is already showing the "Upgrade Q" UI.
+        const isFreeTierLimitUi = store.promptInputStickyCard?.messageId === freeTierLimitSticky.messageId
+        const isUpgradePendingUi = store.promptInputStickyCard?.messageId === upgradePendingSticky.messageId
+        const isPlansAndPricingTab = plansAndPricingTitle === store.tabTitle
+
+        if (mode === 'freetier-limit') {
+            mynahUi.updateStore(tabId, {
+                promptInputStickyCard: freeTierLimitSticky,
+            })
+
+            if (!isFreeTierLimitUi) {
+                // TODO: how to set a warning icon on the user's failed prompt?
```

Comment:
is this a launch blocking TODO?

Reply:
This is a cosmetic question that will need to be done as a followup.